### ISSUE TITLE
Problem when extending Growing Spheres to MNIST

### ISSUE THREAD

**Issue by e-delaney** [thibaultlaugel/growingspheres#2]
Hello. Thanks for sharing the code, it is very helpful. I am trying to apply Growing Spheres to the MNIST dataset. Eventually I will make adjustments to use a Keras model. However, for now I am using a SVM (following your paper). 

The following error appears (during the exploration step); 

```
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-21-366d07308436> in <module>
----> 1 CF.fit(n_in_layer=200, first_radius=1, dicrease_radius=0.8, sparse=False, verbose=True)

~\growingspheres\counterfactuals.py in fit(self, caps, n_in_layer, first_radius, dicrease_radius, sparse, verbose)
     42                 sparse,
     43                 verbose)
---> 44         self.enemy = cf.find_counterfactual()
     45         self.e_star = cf.e_star
     46         self.move = self.enemy - self.obs_to_interprete

~\growingspheres-master\growingspheres\growingspheres.py in find_counterfactual(self)
     55         Finds the decision border then perform projections to make the explanation sparse.
     56         """
---> 57         ennemies_ = self.exploration()
     58         closest_ennemy_ = sorted(ennemies_, 
     59                                  key= lambda x: pairwise_distances(self.obs_to_interprete.reshape(1, -1), x.reshape(1, -1)))[0] 

~\growingspheres-master\growingspheres\growingspheres.py in exploration(self)
     87 
     88             while n_ennemies_ <= 0:
---> 89                 layer = self.ennemies_in_layer_((radius_, radius_ + step_), self.caps, self.n_in_layer)
     90                 n_ennemies_ = layer.shape[0]
     91                 radius_ = radius_ + step_

~\growingspheres-master\growingspheres\growingspheres.py in ennemies_in_layer_(self, segment, caps, n)
    103         Basis for GS: generates a hypersphere layer, labels it with the blackbox and returns the instances that are predicted to belong to the target class.
    104         """
--> 105         layer = generate_inside_ball(self.obs_to_interprete, segment, n)
    106         #cap here: not optimal
    107         if caps != None:

~\growingspheres-master\growingspheres\utils\gs_utils.py in generate_inside_ball(center, segment, n)
     26     d = center.shape[1]
     27     z = np.random.normal(0, 1, (n, d))
---> 28     u = np.random.uniform(segment[0]**d, segment[1]**d, n)
     29     r = u**(1/float(d))
     30     z = np.array([a * b / c for a, b, c in zip(z, r,  norm(z))])

OverflowError: (34, 'Result too large')
```

I've successfully applied Growing Spheres to the Iris dataset so I don't think the number of classes is an issue. I think the issue could be something to do with the dimensionality of the data, and perhaps an improper selection of the parameters (I've tried to vary the radius size and decrease but no luck so far). The shape of the observation instance is (1,784). 

I have also uploaded my notebook [here;](https://github.com/e-delaney/GS_MNIST/blob/main/GS%20simple%20example.ipynb)

Any help would be greatly appreciated. Alternatively, if you can share the code for the MNIST experiment in your paper I might be able to work through it and solve my problem!

Thanks again.

**Comment by thibaultlaugel** [comment]
Hello,
Sorry for the delay in answer and thanks for pointing this out!

So I do not have access to the original code anymore, and I encountered the same problem as you, caused by the dimensionality of MNIST. I am not sure how I originally overcame this issue (sorry about that...), so I spent a few days looking for workarounds or quick-fixes, and the best I could find in a short amount of time was to add a new parameter to the algorithm, called 'layer_shape', which controls the shape of the covered at each sampling iteration. 

layer_shape can take 3 values: 'ring' (default), 'ball' or 'sphere'.

1. 'ring': 
The algorithm initially used for GS to efficiently sample in "rings" (called hyperspherical layers in the paper) required to calculate a^d and b^d (d being the dimension of the problem and a and b the radius values defining the ring). Since this creates an OverflowError in high dimension and I have not been able to quicky find another way of doing this ring sampling efficiently, I suggest using the other sampling shapes for MNIST (and high dimension in general). I will try to spend more time on finding another sampling strategy for high dimension rings.

2. 'ball':
At each iteration, it samples from a ball of larger radiuses. The algorithm used to sample in hyperballs does not require to calculate radius^d, so it works for high dimension. However, this means that the sampling is much less 'efficient', since each iteration re-covers the areas that were already covered by previous iterations.

3. 'sphere':
Sampling on the surface of the sphere. So the space covered is much smaller, but at least it works.

Of course this raises many questions besides the ones that existed already, such as how n_layer should be fixed depending on these shapes and radius values... Hopefully I will work on this 'soon'. 
Note that I have yet to conduct rigorous experiments to evaluate how efficient these new sampling strategies are. 

Hope this helps. I added the MNIST example in the "simple example notebook", and another notebook with an illustration of the sampling strategies. I am closing the issue, do not hesitate to contact me if you still encounter issues.
